variable/include/scipp/variable/variable_factory.h

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp/variable/variable.h"
+
+namespace scipp::variable {
+
+/// Abstract base class for "variable makers", used by VariableFactory to
+/// dynamically create variables with given type.
+class SCIPP_VARIABLE_EXPORT AbstractVariableMaker {
+protected:
+  auto unreachable() const {
+    return std::logic_error("This code path should never be reached");
+  }
+
+public:
+  virtual ~AbstractVariableMaker() = default;
+  virtual bool is_buckets() const = 0;
+  virtual Variable
+  create(const DType elem_dtype, const Dimensions &dims, const bool variances,
+         const std::vector<VariableConstView> &parents) const = 0;
+  virtual DType elem_dtype(const VariableConstView &var) const = 0;
+  virtual bool hasVariances(const VariableConstView &var) const = 0;
+  virtual VariableConstView data(const VariableConstView &) const {
+    throw unreachable();
```

Comment:
So you are doing this `unreachable` trick to avoid having to define these functions (`data()` and `array_params()`) in all the derived classes?

Reply:
Not really a trick, but a sign that the class structure here is slightly odd/smelly: The methods are not used for "normal" dtypes and do not necessarily have a sensible return type. I would have thrown in the derived class as well, but the awkwardness would not have been less.